crates/evm/core/src/backend/fork_type.rs

```diff
@@ -0,0 +1,55 @@
+use std::collections::HashMap;
+
+/// Defines a fork of the type EVM or ZK.
+#[derive(Debug, Clone)]
+pub enum ForkType {
+    Evm,
+    Zk,
+}
+
+impl ForkType {
+    /// Returns true if type is [ForkType::Zk]
+    pub fn is_zk(&self) -> bool {
+        matches!(self, ForkType::Zk)
+    }
+
+    /// Returns true if type is [ForkType::Evm]
+    pub fn is_evm(&self) -> bool {
+        matches!(self, ForkType::Evm)
+    }
+}
+
+/// A cached implementation for retrieving the [ForkType] of a given url.
+#[derive(Default, Debug, Clone)]
+pub struct CachedForkType(HashMap<String, ForkType>);
+
+impl CachedForkType {
+    /// Retrieve the [ForkType] of a url.
+    /// We attempt to query the rpc provider for "zks_L1ChainId" method. If it returns successfully,
+    /// then the chain is [ForkType::Zk], else it's [ForkType::Evm].
+    /// The result is then cached
+    pub fn get(&mut self, fork_url: &str) -> ForkType {
+        if let Some(fork_url_type) = self.0.get(fork_url) {
+            return fork_url_type.clone()
+        }
+
+        let is_zk_url = foundry_common::try_get_http_provider(fork_url)
+            .map(|provider| {
+                let is_zk_url = tokio::runtime::Builder::new_multi_thread()
+                    .enable_all()
+                    .build()
+                    .unwrap()
+                    .block_on(provider.request("zks_L1ChainId", ()))
+                    .map(|_: String| true)
+                    .unwrap_or_default();
+
+                is_zk_url
+            })
+            .unwrap_or_default();
```

Comment:
There's a retry mechanism, so that would take care at least of this partially. If the RPC is really not available then we wouldn't even be able to obtain the information for the fork to actually happen